src/main/java/org/folio/uk/integration/keycloak/KeycloakService.java

```diff
@@ -23,37 +23,113 @@
+
+      return kcUserId;
     }
+
     var kcUser = toKeycloakUser(user, password);
     kcUser.setUserTenantAttr(of(getRealm()));
     log.info("Creating keycloak user: userId = {}", user.getId());
+    return callKeycloak(
+      create(kcUser), () -> buildUsersErrorMessage("Failed to create keycloak user", user.getId()));
+  }
+
+  public void linkIdentityProviderToUser(User user, String kcUserId) {
+    var userId = user.getId();
+    var tenant = getRealm();
+
+    log.info("Linking identity provider to user [userId: {}, kcUserId: {}, tenant: {}, username: {}]", userId,
+      kcUserId, tenant, user.getUsername());
+
+    var userTenantOptional = getUserTenant(userId);
+    if (userTenantOptional.isEmpty() || StringUtils.isEmpty(userTenantOptional.get().getCentralTenantId())) {
+      log.warn("Identity provider cannot be linked to user because userTenant is empty"
+        + " or has empty centralTenantId, [userId: {}, kcUserId: {}, tenant: {}]", userId, kcUserId, tenant);
+      return;
+    }
+
+    var userTenant = userTenantOptional.get();
+    log.debug("Found userTenant for user [userId: {}, kcUserId: {}, tenant: {}, userTenant: {}]", userId,
+      kcUserId, tenant, userTenant);
+
+    // Prevents linking if not as a central OR if as a member tenant
+    var memberTenant = userTenant.getTenantId();
+    if (!tenant.equals(userTenant.getCentralTenantId()) || tenant.equals(memberTenant)) {
+      log.warn("Identity provider cannot be linked to central tenant or member tenant real users, "
+        + "[userId: {}, kcUserId: {}, tenant: {}, memberTenant: {}]", userId, kcUserId, tenant, memberTenant);
+      return;
+    }
+
+    var providerAlias = keycloakFederatedAuthProperties.getIdentityProviderSuffix()
+      .replace("{tenantId}", memberTenant);
+
+    if (isIdentityProviderAlreadyLinked(kcUserId, tenant, providerAlias)) {
+      log.warn("Identity provider is already linked to user [userId: {}, kcUserId: {}, tenant: {}, "
+        + "memberTenant: {}, providerAlias: {}]", userId, kcUserId, tenant, memberTenant, providerAlias);
+      return;
+    }
+
+    var federatedIdentity = createFederatedIdentity(user);
+    callKeycloak(
+      () -> keycloakClient.linkIdentityProviderToUser(tenant, kcUserId, providerAlias, federatedIdentity, getToken()),
+      () -> String.format("Failed to link identity provider to user [userId: %s, kcUserId: %s, tenant: %s, "
+        + "memberTenant: %s, providerAlias: %s]", userId, kcUserId, tenant, memberTenant, providerAlias));
+
+    log.info("Created identity provider for use [userId: {}, kcUserId: {}, tenant: {}, memberTenant: {}, "
+      + "providerAlias: {}, federatedIdentity: {}]", userId, kcUserId, tenant, memberTenant, providerAlias,
+      federatedIdentity);
+  }
+
+  private Optional<UserTenant> getUserTenant(UUID userId) {
+    return userTenantsClient.lookupByUserId(userId).getUserTenants().stream()
+      .filter(Objects::nonNull).findFirst();
```

Comment:
The checks can be simplified here, we always will have user_tenant response here, if it is not ECS mode - the response will have empty result, if ECS mode at lest 1 result, so also can add limit=1 to the request. We can return just centralTenantId or null that will simplify checks in the method linkIdentityProviderToUser().
Also we can intoduce UserTenantService with @Cacheable , we always cache this user-tenant invocation